h/static/scripts/controllers/search-bar-controller.js

```diff
@@ -132,30 +129,31 @@ class SearchBarController extends Controller {
       var hiddenInput = document.createElement('input');
       hiddenInput.type = 'hidden';
 
-      Array.from(this._lozengeContainer.querySelectorAll('.js-lozenge__content')).forEach((loz) => {
-        hiddenInput.value = hiddenInput.value + loz.textContent + ' ';
-      });
-      hiddenInput.value = hiddenInput.value + getTrimmedInputValue();
-
       // When JavaScript isn't enabled this._input is submitted to the server
       // as the q param. With JavaScript we submit hiddenInput instead.
       hiddenInput.name = this._input.name;
       this._input.removeAttribute('name');
 
       this.refs.searchBarForm.appendChild(hiddenInput);
+      return hiddenInput;
     };
 
     /**
-     * Submit the user's search query to the server.
+     * Update the value of the hidden input.
+     *
+     * Update the value of the hidden input based on the contents of any
+     * lozenges and any remaining text in the visible input.
+     *
+     * This should be called whenever a lozenge is added to or removed from
+     * the DOM, and whenever the text in the visible input changes.
      *
-     * We build a search query out of the lozenges plus any text in
-     * this._input. To avoid a potential flash of text if we were to update
-     * this._input with the search terms from the lozenge before submitting it,
-     * we update a hidden input and submit that instead.
      */
-    var submitForm = () => {
-      insertHiddenInput();
-      this.refs.searchBarForm.submit();
+    var updateHiddenInput = () => {
+      let newValue = '';
+      Array.from(this._lozengeContainer.querySelectorAll('.js-lozenge__content')).forEach((loz) => {
+        newValue = newValue + loz.textContent + ' ';
```

Comment:
Yes, if you don't mind I'd like to hang on to <https://github.com/hypothesis/h/pull/4024>, I'll review and merge it after this one is merged. The reason is that this PR is blocking a lot of feature work on activity pages that is due at the end of this sprint.